File: Week2/과제/그래프_탐색/이진_트리_최대_경로_합/Local/Solution_BinaryTree.py
# 시스템 표준 입출력 및 파이썬 관련 시스템 함수들을 제공하는 sys 모듈을 불러옵니다.
import sys


# 이진 트리의 개별 노드를 나타내는 클래스입니다.
# 트리의 각 노드는 하나의 값을 가지고, 왼쪽과 오른쪽 자식 노드를 가리킬 수 있습니다.
class TreeNode:
    # TreeNode 클래스의 생성자 메소드입니다.
    # 파라미터 설명:
    # - val: 노드가 가질 데이터 값 (기본값은 0)
    # - left: 왼쪽 자식 노드를 가리키는 포인터 (기본값은 None)
    # - right: 오른쪽 자식 노드를 가리키는 포인터 (기본값은 None)
    def __init__(self, val=0, left=None, right=None):
        # 노드의 정수 값을 설정합니다.
        self.val = val
        # 왼쪽 자식 노드에 대한 연결 링크를 설정합니다.
        self.left = left
        # 오른쪽 자식 노드에 대한 연결 링크를 설정합니다.
        self.right = right


# 이진 트리 최대 경로 합 문제를 해결하는 클래스입니다.
class Solution:
    # 전체 이진 트리에서 얻을 수 있는 최대 경로 합을 찾아 반환하는 메소드입니다.
    # 파라미터 설명:
    # - root: 탐색할 이진 트리의 최상위 루트 노드 (TreeNode 객체)
    # 반환값: 이진 트리 내부에서 노드들을 연결해 얻을 수 있는 최대 경로 합 (정수)
    def maxPathSum(self, root):
        # 전체 최대 합을 기록하기 위한 변수를 아주 작은 수(마이너스 무한대)로 초기화합니다.
        # 음수 값들로만 가득 찬 트리에서도 올바른 최대값을 찾기 위함입니다.
        self.max_sum = float('-inf')

        # 현재 노드에서 아래(자식) 방향으로 연장해 나갈 때 얻을 수 있는 최대 단방향 합(gain)을 구하는 재귀 함수입니다.
        # 파라미터 설명:
        # - node: 현재 검사하고 있는 노드 (TreeNode 객체)
        # 반환값: 현재 노드를 시작으로 하위 노드 방향으로 이어지는 최대 경로의 단방향 합
        def get_max_gain(node):
            # 만약 현재 검사하는 노드가 비어있다면 (None)
            if not node:
                # 합에 기여할 수 있는 값이 없으므로 0을 반환합니다.
                return 0

            # 왼쪽 자식 노드 방향으로 내려가면서 얻을 수 있는 최대 이득을 구합니다.
            # 재귀적으로 호출하여 하위 노드들의 결과를 가져옵니다.
            # 만약 하위 노드들에서 얻은 최대 이득이 음수라면, 경로에 포함하지 않는 것이 이득이므로 
            # max(..., 0)을 통해 0으로 처리하여 선택하지 않습니다.
            left_gain = max(get_max_gain(node.left), 0)
            
            # 오른쪽 자식 노드 방향으로 내려가면서 얻을 수 있는 최대 이득을 구합니다.
            # 마찬가지로 음수 합이 나온다면 0으로 취급하여 경로에서 배제합니다.
            right_gain = max(get_max_gain(node.right), 0)

            # 현재 노드를 '정점(꺾이는 지점)'으로 삼아 왼쪽과 오른쪽 자식 모두를 연결하는 경로의 합을 구합니다.
            # 즉, [왼쪽 자식 경로 - 현재 노드 - 오른쪽 자식 경로] 형태의 완전한 경로 합입니다.
            current_path_sum = node.val + left_gain + right_gain
            
            # 이 정점(꺾이는 지점)을 통과하는 경로의 합이 기존에 찾아둔 전체 최대 경로 합(self.max_sum)보다 크다면
            # 전체 최대 경로 합을 새로운 더 큰 값으로 갱신해 줍니다.
            self.max_sum = max(self.max_sum, current_path_sum)

            # 이 노드의 상위 부모 노드 입장에서 볼 때, 이 노드를 거쳐 갈 수 있는 방향은 한쪽 방향뿐입니다.
            # (한 노드에서 왼쪽, 오른쪽 자식 양쪽으로 동시에 꺾여 올라갈 수는 없습니다. 경로가 끊어지게 됨)
            # 따라서 부모 노드 방향으로 전달할 때에는 '현재 노드 값 + (왼쪽과 오른쪽 이득 중 더 큰 쪽)'을 선택하여 반환합니다.
            return node.val + max(left_gain, right_gain)

        # 루트 노드부터 탐색을 시작하여 재귀적으로 트리를 분석합니다.
        get_max_gain(root)
        # 계산 완료 후 최종적으로 갱신된 전역 최대 경로 합을 반환합니다.
        return self.max_sum


# 입력을 바탕으로 실제 이진 트리 자료구조를 구축해 주는 함수입니다.
# 레벨 순서 탐색(Queue 이용) 방식으로 트리를 복원합니다.
# 파라미터 설명:
# - nodes: 트리를 구성하는 노드들의 값이 문자열 형태로 담긴 리스트 (예: ["10", "2", "10", "null", "null", "-25"])
# 반환값: 완성된 이진 트리의 최상위 루트 노드 (TreeNode 객체)
def build_tree(nodes):
    # 만약 입력받은 리스트가 비어 있거나, 첫 번째 값(루트)이 비어있거나 "null"이라면
    if not nodes or nodes[0] == "null" or nodes[0] == "":
        # 빈 트리이므로 None을 반환합니다.
        return None

    # 첫 번째 문자열 값을 정수로 변환하여 트리의 최상위 루트 노드를 생성합니다.
    root = TreeNode(int(nodes[0]))
    # 자식 노드를 순차적으로 연결해 주기 위해 대기열 역할을 하는 큐(queue) 리스트를 생성하고 루트를 담습니다.
    queue = [root]
    # 입력 노드 리스트의 다음 인덱스를 가리키는 포인터 변수 i를 1로 지정합니다.
    i = 1

    # 큐에 노드가 있고 리스트에 아직 처리할 데이터가 남아있다면 반복을 진행합니다.
    while queue and i < len(nodes):
        # 큐의 맨 앞에서 자식을 연결해 줄 부모 노드를 하나 꺼냅니다. (pop(0)으로 맨 처음 요소 꺼냄)
        node = queue.pop(0)
        
        # 왼쪽 자식 노드를 처리할 차례입니다.
        if i < len(nodes):
            # 만약 현재 데이터가 문자열 "null"이 아니라면
            if nodes[i] != "null":
                # 정수로 바꾼 값으로 왼쪽 자식 노드를 새로 만듭니다.
                node.left = TreeNode(int(nodes[i]))
                # 새로 만든 왼쪽 자식을 나중에 그 자식들의 부모 역할을 하도록 큐에 대기시킵니다.
                queue.append(node.left)
            # 인덱스를 한 칸 이동합니다.
            i += 1
            
        # 오른쪽 자식 노드를 처리할 차례입니다.
        if i < len(nodes):
            # 만약 현재 데이터가 문자열 "null"이 아니라면
            if nodes[i] != "null":
                # 정수로 바꾼 값으로 오른쪽 자식 노드를 새로 만듭니다.
                node.right = TreeNode(int(nodes[i]))
                # 새로 만든 오른쪽 자식을 큐에 대기시킵니다.
                queue.append(node.right)
            # 인덱스를 한 칸 이동합니다.
            i += 1

    # 완성된 트리의 최상위 루트 노드를 반환합니다.
    return root


# 스크립트가 메인으로 직접 실행될 때 실행되는 구문입니다.
if __name__ == "__main__":
    # 표준 입력으로부터 한 줄을 통째로 읽어와 앞뒤 공백을 제거합니다.
    line = sys.stdin.readline().strip()
    # 입력 내용이 비어있지 않다면
    if line:
        # 입력이 '[1, 2, 3]'과 같이 대괄호로 둘러싸여 있는 경우 대괄호를 잘라냅니다.
        if line.startswith('[') and line.endswith(']'):
            line = line[1:-1]
        # 대괄호를 지운 문자열이 여전히 존재한다면
        if line.strip():
            # 쉼표(,)를 기준으로 쪼갠 후 각 단어의 양끝 공백을 지워 리스트로 만듭니다.
            input_data = [x.strip() for x in line.split(',')]
            # 정제된 리스트를 통해 이진 트리를 구성합니다.
            root = build_tree(input_data)
            # Solution 인스턴스를 생성합니다.
            sol = Solution()
            # 최대 경로 합 알고리즘을 수행하여 그 최종 결과값을 콘솔에 출력합니다.
            print(sol.maxPathSum(root))

File: Week2/과제/그래프_탐색/이진_트리_최대_경로_합/Local/Solution_BinaryTree.cpp
// 최대값, 최소값 계산을 위한 알고리즘(std::max) 라이브러리를 불러옵니다.
#include <algorithm>
// 정수형 타입이 가질 수 있는 한계값(예: INT_MIN 등 최소값)을 담고 있는 라이브러리를 불러옵니다.
#include <climits>
// 표준 입출력을 처리하기 위한 iostream 라이브러리를 불러옵니다.
#include <iostream>
// BFS(너비 우선 탐색)를 활용해 트리를 구축하기 위해 queue 라이브러리를 불러옵니다.
#include <queue>
// 문자열(string)을 처리하기 위한 라이브러리를 불러옵니다.
#include <string>
// 동적 배열(vector)을 다루기 위한 라이브러리를 불러옵니다.
#include <vector>

// std 네임스페이스를 기본으로 사용하여, std:: 접두어 없이 표준 라이브러리 요소를 바로 사용할 수 있게 선언합니다.
using namespace std;

// 이진 트리의 노드 구조를 정의하는 구조체(struct)입니다.
struct TreeNode
{
    // 노드에 담길 정수형 데이터입니다.
    int val;
    // 왼쪽 자식 노드를 가리키는 포인터 주소입니다.
    TreeNode *left;
    // 오른쪽 자식 노드를 가리키는 포인터 주소입니다.
    TreeNode *right;
    
    // 생성자 함수로, 새로운 노드가 만들어질 때 값을 초기화하는 역할을 담당합니다.
    // 파라미터 x를 받아서 val에 대입하고, left와 right 포인터는 NULL(비어있음)로 지정합니다.
    TreeNode(int x) : val(x), left(NULL), right(NULL)
    {
    }
};

// 이진 트리의 최대 경로 합을 계산하는 알고리즘을 담은 클래스입니다.
class Solution
{
  private:
    // 전역적으로 최대 경로 합을 갱신해 나가며 저장할 멤버 변수입니다.
    // 처음에는 정수가 가질 수 있는 가장 작은 값(INT_MIN)으로 초기화하여, 
    // 음수로 이루어진 트리에서도 올바르게 최대값 연산이 되도록 합니다.
    int max_sum = INT_MIN;

  public:
    // 최대 경로 합 계산을 시작하고 최종 결과를 반환하는 메소드입니다.
    // 파라미터 설명:
    // - root: 탐색할 이진 트리의 최상위 노드 포인터 (TreeNode*)
    // 반환값: 전체 이진 트리에서 얻을 수 있는 최대 경로의 누적 합 (int)
    int maxPathSum(TreeNode *root)
    {
        // 재귀 탐색 함수를 수행하여 트리 전체를 검사하고 max_sum 변수를 업데이트합니다.
        get_max_gain(root);
        // 계산이 완료된 후 최종적으로 저장된 최대 경로 합을 반환합니다.
        return max_sum;
    }

    // 현재 노드를 기준으로 아래 방향(자식)으로 뻗어나가는 경로의 최대값(gain)을 구하는 재귀 함수입니다.
    // 파라미터 설명:
    // - node: 현재 탐색 중인 노드의 포인터 (TreeNode*)
    // 반환값: 현재 노드를 포함하여 아래 자식 방향으로 연결했을 때 얻을 수 있는 최대 단방향 경로 합
    int get_max_gain(TreeNode *node)
    {
        // 만약 가리키는 노드가 없다면 (빈 트리 또는 자식 노드가 없는 상태)
        if (node == NULL)
        {
            // 합에 기여할 수 있는 값이 없으므로 0을 반환합니다.
            return 0;
        }

        // 왼쪽 자식 노드 방향으로 내려가면서 얻을 수 있는 최대 이득(gain)을 구합니다.
        // 재귀 호출을 수행하고, 음수 값이 나온다면 경로에서 제외하는 것이 최대값을 유지하는 데 이득이므로 
        // 0보다 작은 결과는 max(..., 0) 처리를 통해 버립니다.
        int left_gain = max(get_max_gain(node->left), 0);
        
        // 오른쪽 자식 노드 방향으로 내려가며 얻을 수 있는 최대 이득을 구합니다.
        // 마찬가지로 0 미만의 음수가 나온다면 0으로 취급하여 배제합니다.
        int right_gain = max(get_max_gain(node->right), 0);

        // 현재 노드를 경로의 꺾이는 정점으로 간주하고, [왼쪽 자식 - 현재 노드 - 오른쪽 자식]을 하나로 연결한 합을 계산합니다.
        int current_path_sum = node->val + left_gain + right_gain;
        
        // 이 꺾이는 경로의 합이 지금까지 발견된 전체 최대 경로 합(max_sum)보다 크다면 업데이트합니다.
        max_sum = max(max_sum, current_path_sum);

        // 부모 노드의 관점에서는, 이 노드를 거쳐 갈 때 한쪽 방향(왼쪽 혹은 오른쪽)으로만 갈 수 있으므로,
        // '현재 노드의 값 + (왼쪽과 오른쪽 이득 중 더 큰 쪽)'을 선택하여 반환합니다.
        return node->val + max(left_gain, right_gain);
    }
};

// 입력 데이터를 받아 실제 이진 트리 형태로 복원하고 구축해 주는 함수입니다.
// 파라미터 설명:
// - nodes: 노드의 값들이 순서대로 담긴 문자열 형태의 벡터 배열 (예: {"10", "2", "10", "null"})
// 반환값: 구성이 완료된 트리의 최상위 루트 노드 포인터 (TreeNode*)
TreeNode *buildTree(vector<string> nodes)
{
    // 입력 리스트가 비어 있거나, 첫 번째 값이 비어있는 "null" 노드라면
    if (nodes.empty() || nodes[0] == "null")
    {
        // 빈 트리이므로 NULL을 반환합니다.
        return NULL;
    }
    
    // 첫 번째 노드 데이터를 정수로 변환(stoi 사용)하여 루트 노드를 만듭니다.
    TreeNode *root = new TreeNode(stoi(nodes[0]));
    // 자식 노드들을 차례대로 이어주기 위해 큐 자료구조를 생성하고 생성한 루트를 집어넣습니다.
    queue<TreeNode *> q;
    q.push(root);
    // 입력 리스트의 다음 가리킬 요소 번호(인덱스)인 i를 1로 지정합니다.
    int i = 1;
    
    // 큐가 비어있지 않고, 리스트의 끝에 도달할 때까지 노드들을 하나씩 구성합니다.
    while (!q.empty() && i < nodes.size())
    {
        // 자식을 이어줄 기준이 될 부모 노드를 큐의 맨 앞에서 꺼내옵니다.
        TreeNode *curr = q.front();
        q.pop();
        
        // 왼쪽 자식 노드를 처리합니다.
        if (nodes[i] != "null")
        {
            // 정수 값으로 변환하여 왼쪽 자식 노드를 동적으로 할당하여 생성합니다.
            curr->left = new TreeNode(stoi(nodes[i]));
            // 만들어진 왼쪽 자식을 나중에 그 아래 자식들도 연결할 수 있도록 큐에 대기시킵니다.
            q.push(curr->left);
        }
        // 인덱스를 한 칸 오른쪽으로 이동합니다.
        i++;

        // 오른쪽 자식 노드를 처리합니다.
        if (i < nodes.size() && nodes[i] != "null")
        {
            // 정수 값으로 변환하여 오른쪽 자식 노드를 생성합니다.
            curr->right = new TreeNode(stoi(nodes[i]));
            // 오른쪽 자식 노드도 아래 자식 연결 처리를 위해 큐에 넣습니다.
            q.push(curr->right);
        }
        // 인덱스를 한 칸 오른쪽으로 이동합니다.
        i++;
    }

    // 트리 조립이 완료되었으므로 루트 노드의 주소(포인터)를 반환합니다.
    return root;
}

// 프로그램의 메인 실행부입니다.
int main()
{
    string line;
    vector<string> nodes;
    // 콘솔 입력을 통해 한 줄을 통째로 받아들입니다.
    if (getline(cin, line))
    {
        // 문자열 앞뒤의 의미 없는 공백 문자들을 찾아 제거합니다(Trim).
        line.erase(0, line.find_first_not_of(" \t\r\n"));
        if (!line.empty())
        {
            line.erase(line.find_last_not_of(" \t\r\n") + 1);
        }

        // 입력받은 문자열이 대괄호 '[' 로 시작하면 첫 문자를 제외합니다.
        if (!line.empty() && line.front() == '[')
        {
            line = line.substr(1);
        }
        // 입력받은 문자열이 대괄호 ']' 로 끝나면 마지막 문자를 제외합니다.
        if (!line.empty() && line.back() == ']')
        {
            line = line.substr(0, line.size() - 1);
        }

        // 대괄호를 걷어낸 문자열을 쉼표(,) 기준으로 토큰화(쪼개기)하여 벡터 배열에 넣습니다.
        if (!line.empty())
        {
            size_t pos = 0;
            while ((pos = line.find(',')) != string::npos)
            {
                string token = line.substr(0, pos);
                // 쉼표 앞 단어의 공백 제거 처리
                token.erase(0, token.find_first_not_of(" \t\r\n"));
                if (token.find_last_not_of(" \t\r\n") != string::npos)
                {
                    token.erase(token.find_last_not_of(" \t\r\n") + 1);
                }
                else
                {
                    token = "";
                }
                // 쪼갠 요소를 노드 목록 벡터에 추가합니다.
                nodes.push_back(token);
                // 방금 가공한 부분과 쉼표를 원본 문자열에서 지웁니다.
                line.erase(0, pos + 1);
            }
            
            // 쉼표 뒤에 마지막으로 남은 단어를 처리하여 추가합니다.
            line.erase(0, line.find_first_not_of(" \t\r\n"));
            if (line.find_last_not_of(" \t\r\n") != string::npos)
            {
                line.erase(line.find_last_not_of(" \t\r\n") + 1);
            }
            else
            {
                line = "";
            }
            if (!line.empty())
            {
                nodes.push_back(line);
            }
        }
    }

    // 입력받은 노드 명단을 기반으로 트리 구조를 buildTree 함수로 완성합니다.
    TreeNode *root = buildTree(nodes);
    // Solution 객체를 선언합니다.
    Solution sol;
    // 최대 경로 합 연산을 실행하고 결과를 화면에 출력합니다.
    cout << sol.maxPathSum(root) << endl;

    // 프로그램이 완벽하게 마무리되었음을 나타냅니다.
    return 0;
}
